Add nexusformat examples

rayosborn opened the pull request:
* Adds additional scripts to `manual/examples/h5py` that use the [nexusformat](https://github.com/nexpy/nexusformat) package to produce equivalent output to the existing scripts that use `h5py` directly.
* Modifies the Sphinx files to allow the scripts to be displayed in parallel using tabs created by the [sphinx-tabs](https://github.com/executablebooks/sphinx-tabs) package.

This PR addresses #1117.

rayosborn commented:
I presume that the Github Actions have been modified since I made my fork, so that the `env_workflow.yml` file no longer exists in the main branch. The `sphinx-tabs` package needs to be added to the workflow dependencies somewhere else.

woutdenolf commented:
> I presume that the Github Actions have been modified since I made my fork, so that the `env_workflow.yml` file no longer exists in the main branch. The `sphinx-tabs` package needs to be added to the workflow dependencies somewhere else.
> 
> Edit: Unless the dependencies are now handled by `requirements.txt`. I added `sphinx-tabs` there already.

Yes the workflow is now using the `requirements.txt` file so add all dependencies you need there. The `.github/env-workflow.yml` no longer exists.

rayosborn commented:
Thanks, @woutdenolf. I agree entirely. I had intended to remove the `nexusformat` directory, but forgot when I wrote the PR.

rayosborn commented:
I have renamed the `h5py` directory to `python`, as suggested, and updated the references I was able to identify. I'm not entirely sure of the Sphinx rules for capitalizing references - it looks as if they are case-insensitive, but I attempted to follow the previous capitalization when replacing, e.g., `Example-H5py` with `Example-Python`.

rayosborn commented:
@prjemian, I would appreciate a little help in resolving the conflicts in `requirements.txt`. When I first issued the PR, I think the only change I made was to add `sphinx-tabs`. Obviously, the file has been updated since. Is it correct that `pyRestTable` is no longer needed? I can't recall adding it myself (it was a while back now), but I wanted to be sure before uploading a new version without it.

prjemian commented:
@rayosborn - That's right, the `pyRestTable` requirement was removed in #1122 (as noted in the [v2022.06 release notes](https://github.com/nexusformat/definitions/wiki/releasenotes__v2022.06)).  That package is no longer used in this repository.

rayosborn commented:
@prjemian - Thanks for your help.

prjemian commented:
Sphinx is stopping with this exception:

```
Warning, treated as error:
/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/examples/python/external_example_write/index.rst:4:duplicate label h5py-example-external-links, other instance in /home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/examples/h5py/external_example_write/index.rst
make[1]: *** [Makefile:62: html] Error 2
```

The second file is not in the source code: `./manual/source/examples/h5py/external_example_write/index.rst`
What builds this file?

prjemian commented:
CI failed on a different line of the Sphinx Makefile:

```
Warning, treated as error:
/home/runner/work/definitions/definitions/build/manual/source/examples/python/index.rst:19:Too many autonumbered footnote references: only 3 corresponding footnotes available.
make: *** [Makefile:62: html] Error 2
```

prjemian commented:
After correcting the footnote problem in my local build, Sphinx stops with this error:

```
Warning, treated as error:
/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/examples/python/simple_example_basic/index.rst:94:Explicit markup ends without a blank line; unexpected unindent.
make[1]: *** [Makefile:62: html] Error 2
```

which is generated from this code snippet in file `./manual/source/examples/python/simple_example_basic/index.rst`:

```rst
    .. rubric:: *simple_example_basic_read.py*: Read a NeXus HDF5 file using Python
    p
    .. _Example-Python-Reader:

```

Line 94 is the line after the `.. rubric::`, the one with the `p` all by itself.

rayosborn commented on manual/source/examples/python/index.rst:
```diff
@@ -1,22 +1,27 @@
+``nexusformat`` [#]_ and ``h5py`` [#]_ packages.  The ``nexusformat`` package provides a simplified syntax for 
+reading and writing NeXus-compliant files by automatically handling some of the features required by the 
+NeXus standard, such as the attributes that define group classes and plottable data. However, it also uses 
+the ``h5py`` package [#]_ to read/write the HDF5 files on disk. We provide tabbed examples showing 
```
I think I have fixed it in my fork.

prjemian commented:
I get this error with your most recent commit:

```
Warning, treated as error:
/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/examples/python/simple_example_basic/index.rst:107:Include file '/home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/examples/python/simple_example_basic/simple_example_basic_read.py.py' not found or reading it failed
make[1]: *** [Makefile:62: html] Error 2
```

prjemian commented:
`.py.py`  Too many `.py`?

prjemian commented:
That's line 108.

prjemian commented:
Same situation on line 55 of file `manual/source/examples/python/simple_example_write1/index.rst`

prjemian commented:
view of my local build with my edit (just before your most recent changes):
![screen](https://user-images.githubusercontent.com/2279984/190695816-89bb1467-12a8-42a9-a6e2-398e324003aa.png)

prjemian commented:
Same thing with your latest revisions.  Good.

prjemian commented:
Do you want to add any download links to the `downloads` at the bottom of the simple_write_example page?

rayosborn commented:
I had not thought of the downloads. That's a great suggestion.

rayosborn commented:
I had to shift a couple of the columns to make space for the download links. I hope that's what Sphinx expects.

prjemian commented:
Sphinx has a problem.

```
===========================================        =============================================
file                                               description
===========================================        =============================================
:download:`external_angles_h5dump.txt`             *h5dump* analysis of *external_angles.hdf5*
:download:`external_angles.hdf5`                   HDF5 file written by *external_example_write*
:download:`external_angles_structure.txt`          *punx tree* analysis of *external_angles.hdf5*
:download:`external_counts_h5dump.txt`             *h5dump* analysis of *external_counts.hdf5*
:download:`external_counts.hdf5`                   HDF5 file written by *external_example_write*
:download:`external_counts_structure.txt`          *punx tree* analysis of *external_counts.hdf5*
:download:`external_example_write.py`              h5py code to write external linking examples
:download:`nexusformat/external_example_write.py`  nexusformat code to write external linking examples
:download:`external_master_h5dump.txt`             *h5dump* analysis of *external_master.hdf5*
:download:`external_master.hdf5`                   NeXus file written by *external_example_write*
:download:`external_master_structure.txt`          *punx tree* analysis of *external_master.hdf5*
===========================================        =============================================
make[1]: *** [Makefile:62: html] Error 2
```

The `===` length must match the content in all but the last column. Make the file column `===` longer.  It's a tedious thing.

prjemian commented:
[This comment](https://github.com/nexusformat/definitions/pull/1198#issuecomment-1249629816) belongs to PR #1145.

prjemian commented:
tables need to be corrected in:  `manual/source/examples/python/plotting/index.rst` and `manual/source/examples/python/external_example_write/index.rst`, then sphinx builds with no errors

prjemian commented:
```rst
==================================================  =============================================
file                                                description
==================================================  =============================================
:download:`reader_attributes_trail.py`              h5py code to read NeXus HDF5 file and find plottable data
:download:`nexusformat/reader_attributes_trail.py`  nexusformat code to read  NeXus HDF5 file and find plottable data
==================================================  =============================================
```

```rst
================================================== =============================================
file                                               description*
================================================== =============================================
:download:`external_angles_h5dump.txt`             *h5dump* analysis of *external_angles.hdf5*
:download:`external_angles.hdf5`                   HDF5 file written by *external_example_write*
:download:`external_angles_structure.txt`          *punx tree* analysis of *external_angles.hdf5*
:download:`external_counts_h5dump.txt`             *h5dump* analysis of *external_counts.hdf5*
:download:`external_counts.hdf5`                   HDF5 file written by *external_example_write*
:download:`external_counts_structure.txt`          *punx tree* analysis of *external_counts.hdf5*
:download:`external_example_write.py`              h5py code to write external linking examples
:download:`nexusformat/external_example_write.py`  nexusformat code to write external linking examples
:download:`external_master_h5dump.txt`             *h5dump* analysis of *external_master.hdf5*
:download:`external_master.hdf5`                   NeXus file written by *external_example_write*
:download:`external_master_structure.txt`          *punx tree* analysis of *external_master.hdf5*
================================================== =============================================
```

Also, as a matter of style, please end the files with a blank line.  Should content ever get appended to the file, it makes the change history much cleaner.

prjemian commented:
What's the idea for the trailing `*` in the `description*` column?  Can it be removed?

rayosborn commented:
I don't think I have the `description*` in any of my files. I agree it shouldn't be there.

prjemian commented:
Seemed to be in my local.  VSCode editor on my local workstation latches onto the window focus and does not release until other windows are first clocked.  A constant nuisance.

prjemian commented:
That builds on my local workstation.  And also here in the CI.  I'll review later today.

prjemian reviewed (approved):
This looks great now!  Thanks for the contribution and the new web site feature (tabbed views for different languages).  You can merge now, or wait until next week for @woutdenolf to comment.

rayosborn commented:
Great. I didn’t know I had the privilege to merge it myself, but I will see if @woutdenolf has any comments before merging it later on Monday.

woutdenolf reviewed (dismissed):
Very nice @rayosborn ! 

I only have one request: add [this](https://github.com/nexusformat/exampledata/blob/-/autogenerated_examples/nxdl/python_scripts/nexusformat/README.md) link which got lost when removing `manual/source/examples/nexusformat/index.rst`. For example like this

```rst
Code examples
=============

.. toctree::
	:maxdepth: 1

	simple_example_basic/index
	simple_example_write1/index
	simple_example_write2/index
	external_example_write/index
	plotting/index

* `Write examples for different NeXus classes <https://github.com/nexusformat/exampledata/blob/-/autogenerated_examples/nxdl/python_scripts/h5py/README.md>`_
```

```rst
Code examples
=============

.. toctree::
	:maxdepth: 1

	simple_example_basic/index
	simple_example_write1/index
	simple_example_write2/index
	external_example_write/index
	plotting/index

External code examples
======================

* `Write examples for different NeXus classes with h5py <https://github.com/nexusformat/exampledata/blob/-/autogenerated_examples/nxdl/python_scripts/h5py/README.md>`_
* `Write examples for different NeXus classes with nexusformat <https://github.com/nexusformat/exampledata/blob/-/autogenerated_examples/nxdl/python_scripts/nexusformat/README.md>`_
```

rayosborn commented:
@woutdenolf, thanks for catching that. I think I have made the requested change, but please let me know if I missed something.

woutdenolf reviewed (approved):
Thanks @rayosborn ! LGTM

prjemian commented:
Yes, there is another step which is the release of the new version of the NXDL.  This issue and PR are part of the 2023.06 milestone, looking at release next summer.  The new documentation is published with a new release tag.

rayosborn commented:
There must surely be a faster way to release this. Are all documentation updates going to wait a year, even if it's to fix a typo?

prjemian commented:
I believe we can force the documentation to publish without tagging a new
release. (Recently, a new rule imposed the limitation that the docs only
publish with a new tag.) I'll check up on that.

On Tue, Sep 20, 2022, 5:07 PM Ray Osborn ***@***.***> wrote:

> There must surely be a faster way to release this. Are all documentation
> updates going to wait a year, even if it's to fix a typo?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nexusformat/definitions/pull/1145#issuecomment-1252499682>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARMUMBSV7YTOOMEVIVTQIDV7HHMDANCNFSM52GAFT7A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

rayosborn commented:
I'd appreciate that. I presume we would also want to update the documentation to reflect recent NIAC decisions, so waiting a year could be problematic. If you think this needs a NIAC discussion, let me know but the impression I got was that the NIAC was fairly relaxed about changing documentation that didn't involve any policy changes.

woutdenolf commented:
The docs already contain changes to the NeXus standard since 2022.07 so we should make a proper 2022.09 release imo.

rayosborn commented:
@prjemian, thank you for expediting this. I think the results look great. I hope it's helpful to newcomers.
